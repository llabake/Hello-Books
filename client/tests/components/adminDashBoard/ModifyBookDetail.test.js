import React from 'react';

import {
  ModifyBookDetail,
  mapDispatchToProps,
  mapStateToProps
} from '../../../components/adminDashBoard/ModifyBookDetail';

import adminData from '../../mocks/userData'

function setup() {
  const props = {
    dispatch: jest.fn(),
    user: adminData.admin,
    errors: {},
    loading: false,
    updateBook: jest.fn(),
    checkIsbnExist: jest.fn().mockImplementation(() => Promise.resolve(1)),
    handleChange: jest.fn(),
    handleSubmit: jest.fn(),
    handleBlur: jest.fn(),
    book: {
      id: 3,
      isbn: 526875,
      title: 'The Secret life of Baba Segi\'s wives',
      author: 'Lola Shoneyin',
      description: 'The secret of the lorax features a man selling air',
      image: 'book image',
      quantity: 23,
      updatedAt: '2018-02-20T35:20:01.852Z',
      createdAt: '2018-02-20T35:20:01.852Z'
    }
  };

  const state = {
    title: '',
    author: '',
    publishedYear: '',
    isbn: '',
    quantity: '',
    description: '',
    aboutAuthor: '',
    image: '',
    errors: {},
    isValid: false,
  };

  const event = {
    preventDefault: jest.fn(),
    target: {
      name: 'title',
      value: 'title'
    }
  };
  const wrapper = shallow(<ModifyBookDetail {...props} {...state} />);

  return {
    wrapper,
    props,
    state,
    event
  };
}


describe('Modify Book page', () => {
  test('should render with the right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  })

  test('should change state for title', () => {
    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      loading: false,
      updateBook: jest.fn(),
      checkIsbnExist: jest.fn().mockImplementation(() => Promise.resolve(1)),
      handleChange: jest.fn(),
      handleSubmit: jest.fn(),
      handleBlur: jest.fn(),
      book: {}
    };
    const { event, state } = setup();
    const wrapper = shallow(<ModifyBookDetail {...props} /> )
    wrapper.find('#title').simulate('change', event);
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.title).toBe('title');
  });

  test('should call handle change', () => {
    const { state } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {
        title: 'another title'
      }
    }
    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      loading: false,
      updateBook: jest.fn(),
      checkIsbnExist: jest.fn().mockImplementation(() => Promise.resolve(1)),
      handleChange: jest.fn(),
      handleSubmit: jest.fn(),
      handleBlur: jest.fn(),
      book: {}
    };
    const wrapper = shallow(<ModifyBookDetail {...props} /> )
    wrapper.setState({
      title: 'another title',
    });
    props.updateBook.mockReset();
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.title).toBe('another title');
  });

  test('should call book exist to check if book exists before modifying book', () => {
    const { wrapper, props, state } = setup();
    const handleBlurSpy = jest.spyOn(ModifyBookDetail.prototype, 'handleBlur')
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'isbn',
        value: '9871645327878'
      }
    }
    wrapper.setState({
      isbn: '9871645327878',
      // isbnExist: {
      //   isbn: 'isbn error'
      // }
    });
    wrapper.instance().handleBlur(event)
    expect(wrapper.instance().state.isbn).toBe('9871645327878');
    expect(props.checkIsbnExist).toHaveBeenCalled()
  });


  test('should handle form submission when the form inputs are valid', () => {
    const { state } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {}
    }

    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      loading: false,
      updateBook: jest.fn(),
      checkIsbnExist: jest.fn().mockImplementation(() => Promise.resolve(1)),
      handleChange: jest.fn(),
      handleSubmit: jest.fn(),
      handleBlur: jest.fn(),
      book: {}
    };
    const wrapper = shallow(<ModifyBookDetail {...props} {...state} /> )
    wrapper.setState({
      title: 'title',
      author: 'author'
    });
    props.updateBook.mockReset();
    wrapper.instance().handleSubmit(event)
    expect(wrapper.instance().state.title).toBe('title');
    expect(wrapper.instance().state.author).toBe('author');
    // expect(props.updateBook).toHaveBeenCalled()
  });

  test('should dispatch needed actions', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('checkIsbnExist')
    expect(mapDispatchToProps(dispatch)).toHaveProperty('updateBook')
    const { checkIsbnExist, updateBook } = mapDispatchToProps(dispatch);
    checkIsbnExist();
    updateBook();
    expect(dispatch).toHaveBeenCalled();
  })

  test('should return an object with props required by component from store', () => {
    const store = {
      errors: {},
    };
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('errors');
  });
})