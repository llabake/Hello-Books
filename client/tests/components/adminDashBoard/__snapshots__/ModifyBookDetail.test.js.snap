// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modify Book page should render with the right amount of elements 1`] = `
<div>
  <div
    id="banner"
  >
    <div
      className="container form-style"
    >
      <div
        className="row"
      >
        <div
          className="row signup-head"
        >
          <div
            className="col s12"
          >
            <h5
              className="center-align"
            >
               Edit Book 
            </h5>
          </div>
        </div>
        <div
          className="col s12 "
        >
          <div
            className="row "
          >
            <form
              className="col s12 signup"
              method="post"
              onSubmit={[Function]}
            >
              <TextInput
                icon="title"
                id="title"
                name="title"
                onChange={[Function]}
                placeholder="Title"
                type="text"
                value="The Secret life of Baba Segi's wives"
              />
              <TextInput
                icon="account_circle"
                id="author"
                name="author"
                onChange={[Function]}
                placeholder="Author"
                type="text"
                value="Lola Shoneyin"
              />
              <TextInput
                icon="date_range"
                id="publishedYear"
                name="publishedYear"
                onChange={[Function]}
                placeholder="Published Year"
                type="text"
              />
              <TextInput
                icon="lock"
                id="isbn"
                name="isbn"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="ISBN"
                type="text"
                value={526875}
              />
              <TextInput
                icon="date_range"
                id="quantity"
                name="quantity"
                onChange={[Function]}
                placeholder="Quantity"
                type="number"
                value={23}
              />
              <TextAreaInput
                icon="short_text"
                id="description"
                name="description"
                onChange={[Function]}
                placeholder="Description"
                value="The secret of the lorax features a man selling air"
              />
              <TextAreaInput
                icon="short_text"
                id="aboutAuthor"
                name="aboutAuthor"
                onChange={[Function]}
                placeholder="About the Author"
              />
              <div
                className="file-field input-field"
                id="image"
                name="image"
                style={
                  Object {
                    "marginTop": "10em",
                  }
                }
                value="book image"
              >
                <div
                  className="btn"
                  style={
                    Object {
                      "fontSize": "13px",
                      "marginTop": "34px",
                      "width": "40%",
                    }
                  }
                >
                  <span>
                    Upload Image
                  </span>
                  <input
                    accept="image/*"
                    onChange={[Function]}
                    style={
                      Object {
                        "marginTop": "74px",
                      }
                    }
                    type="file"
                  />
                </div>
                <div
                  className="file-path-wrapper"
                >
                  <input
                    className="file-path validate"
                    style={
                      Object {
                        "marginTop": "29px",
                      }
                    }
                    type="text"
                  />
                </div>
              </div>
              <button
                className="waves-effect waves-light btn"
                disabled={true}
                type="submit"
              >
                Update Book
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
